src/net_setup.c

```diff
@@ -27,326 +27,268 @@
-		return true;
+/// Read device class, blacklist status and submesh from a host config file. Used at startup when reading all host config files.
+bool node_read_partial(meshlink_handle_t *mesh, node_t *n) {
+	config_t config;
+	packmsg_input_t in;
+
+	if(!node_get_config(mesh, n, &config, &in)) {
+		return false;
 	}
 
-	char *p;
+	char *submesh_name = packmsg_get_str_dup(&in);
 
-	if(!c->config_tree) {
-		init_configuration(&c->config_tree);
+	if(!strcmp(submesh_name, CORE_MESH)) {
+		free(submesh_name);
+		n->submesh = NULL;
+	} else {
+		n->submesh = lookup_or_create_submesh(mesh, submesh_name);
+		free(submesh_name);
 
-		if(!read_host_config(mesh, c->config_tree, c->name)) {
+		if(!n->submesh) {
+			config_free(&config);
 			return false;
 		}
 	}
 
-	/* First, check for simple ECDSAPublicKey statement */
+	int32_t devclass = packmsg_get_int32(&in);
+	bool blacklisted = packmsg_get_bool(&in);
+	config_free(&config);
 
-	if(get_config_string(lookup_config(c->config_tree, "ECDSAPublicKey"), &p)) {
-		c->ecdsa = ecdsa_set_base64_public_key(p);
-		free(p);
-		return c->ecdsa;
+	if(!packmsg_input_ok(&in) || devclass < 0 || devclass > _DEV_CLASS_MAX) {
+		return false;
 	}
 
-	return false;
+	n->devclass = devclass;
+	n->status.blacklisted = blacklisted;
+	return true;
 }
 
-bool read_ecdsa_private_key(meshlink_handle_t *mesh) {
-	FILE *fp;
-	char filename[PATH_MAX];
-
-	snprintf(filename, PATH_MAX, "%s" SLASH "ecdsa_key.priv", mesh->confbase);
-	fp = fopen(filename, "rb");
-
-	if(!fp) {
-		logger(mesh, MESHLINK_ERROR, "Error reading ECDSA private key file: %s", strerror(errno));
-		return false;
+/// Read the public key from a host config file. Used whenever we need to start an SPTPS session.
+bool node_read_public_key(meshlink_handle_t *mesh, node_t *n) {
+	if(ecdsa_active(n->ecdsa)) {
+		return true;
 	}
 
-	mesh->self->connection->ecdsa = ecdsa_read_pem_private_key(fp);
-	fclose(fp);
+	config_t config;
+	packmsg_input_t in;
 
-	if(!mesh->self->connection->ecdsa) {
-		logger(mesh, MESHLINK_ERROR, "Reading ECDSA private key file failed: %s", strerror(errno));
+	if(!node_get_config(mesh, n, &config, &in)) {
+		return false;
 	}
 
-	return mesh->self->connection->ecdsa;
-}
+	packmsg_skip_element(&in); /* submesh */
+	packmsg_get_int32(&in); /* devclass */
+	packmsg_get_bool(&in); /* blacklisted */
 
-static bool read_invitation_key(meshlink_handle_t *mesh) {
-	FILE *fp;
-	char filename[PATH_MAX];
+	const void *key;
+	uint32_t len = packmsg_get_bin_raw(&in, &key);
 
-	if(mesh->invitation_key) {
-		ecdsa_free(mesh->invitation_key);
-		mesh->invitation_key = NULL;
+	if(len != 32) {
+		config_free(&config);
+		return false;
 	}
 
-	snprintf(filename, PATH_MAX, "%s" SLASH "invitations" SLASH "ecdsa_key.priv", mesh->confbase);
+	n->ecdsa = ecdsa_set_public_key(key);
 
-	fp = fopen(filename, "rb");
+	// While we are at it, read known address information
+	if(!n->canonical_address) {
+		n->canonical_address = packmsg_get_str_dup(&in);
```

Comment:
In `meshlink_close()` there is no free performed operation of `n->canonical_address`

Reply:
Indeed. This should be done in `free_node()`.